### DISCUSSION TITLE
THM Elbows

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#22787] [Q&A General] [answered]
Hi Folks

Quick one, the documentation for Elbow1Phase seems to be incorrect, or at the very least confusing. Direction - Direction of axis from start position to end position (no need to normalize). Maybe I'm not clear about the co-ordinate system, usually I've seen tori referred to as the direction being the vector orthogonal to broadest part of the ring, i.e.

![image](https://user-images.githubusercontent.com/2439754/203753042-4f2ccfff-1352-40fa-b956-43ab39f871a8.png)

The image above would be a torus with a direction of [0,0,1] or equally [0,0,-1]. For the system that I've built, see the image below, I have two pipes connected to the elbow (with VolumeJunction1Phase) 

![image](https://user-images.githubusercontent.com/2439754/203754609-ee1a1f42-39c5-4c25-9d3a-b5a29bbf5492.png)

As can be seen, the elbows are in the y-z plane, and thus (I would) described as a x-torus or a torus with a direction of [1,0,0], however when I input that into MOOSE, the network looks as thus;

![image](https://user-images.githubusercontent.com/2439754/203757496-04168ac8-ec72-4401-80ef-444682ad1d83.png)

Which clearly isn't right, the values are fine, the geometry is clearly wrong, if I change instead the direction to be that of a z torus, i.e. [0,0,1], then all is well;
![image](https://user-images.githubusercontent.com/2439754/203757806-37e9fabf-02b8-44ce-9b54-8b0ee079d379.png)

So, I guess I'm confused why the direction should be what it is, what is the underlying coordinate system that makes this make sense?

Also, why can't I use regular connections between straight pipes and elbow bends, given that they are indeed parallel where they meet? Am I confused as what parallel means in this context?

**Comment by GiudGiud** [comment]
@licharlot @joshuahansel

**Comment by licharlot** [comment, marked as answer]
Hi, 
I don't know the exact reason for direction being defined as it is. `direction` describes the direction at the start of the elbow. 

With the current logic that is implemented in THM, you cannot use the `OneToOneJunction1Phase` or `JunctionParallelChannels1Phase` to connect the elbows to a straight pipe because of the discretization of the elbow. The first element is not exactly parallel to that direction.  So you need to use the `VolumeJunction1Phase` component.

**Reply by makeclean** [reply]
Perfect, thanks that makes at least some sense - one suggestion is to update the documentation to reflect that, along with a note that position in this context means the origin of the circle that would make the elbow.

**Reply by joshuahansel** [reply]
We use the direction definition that we do because it could be used universally. It sounds like the definition you mention only pertains to tori. I agree that we should clarify the description - it makes perfect sense for straight pipes, but for curved pipes, it's wrong. I propose "Direction of axis from start position to end position (no need to normalize). For curved channels, it is the inward tangent vector at the start position." Does that sound clear?

We should also at least add an option to make those first elements straight.

**Reply by makeclean** [reply]
Yeah much more clear

**Comment by makeclean** [comment]
Sorry to re-invigorate this one, I'm still quite confused about elbows and the 'inward directed tangent at the start position' - consider the below diagram

![image](https://user-images.githubusercontent.com/2439754/210184921-f61b8282-aea8-4dae-b8dc-1456095ee86c.png)

I've drawn in red, what I consider given our discussions about what the inward directed tangent from the start position. Thus I would say following the flow direction, the directions for the elbows would be [0 -1 0], [0 0 1], [0 1 0], and [0 0 1]. However MOOSE does not agree, it wants them to be Z directed. What key piece of information am I missing? @joshuahansel

**Reply by joshuahansel** [reply]
Your red arrows are not tangent - they are orthogonal to your curve. The black arrow is the correct vector for your first segment.

**Reply by makeclean** [reply]
So these would be right?

![image](https://user-images.githubusercontent.com/2439754/210409616-0b37715e-85f2-472f-b17d-546d7bbbd1b9.png)

MOOSE still wants them to be all [0,0,1] though, that's what I'm really not getting.

I've build half the network in MOOSE;
![image](https://user-images.githubusercontent.com/2439754/210410416-255d6d54-7e5a-4f7d-9f5f-83c40acda3ff.png)
But it needs those tori to be both z-directed, which is inconsistent with above.

**Reply by joshuahansel** [reply]
Your red arrows are correct. What do you mean by "MOOSE wants/needs"? What is requiring the z direction?

**Reply by makeclean** [reply]
Well to properly create the pipe network shown above (red-white-green-blue) I have to use the following (shortened) components with directions
```
[Components]
[pipe_1]
....
type=FlowChannel1Phase
orientation="0.0 0.0 1.0"
...
[]
[pipe_2]
...
type=ElbowPipe1Phase
orientation="0.0 0.0 1.0"
...
[]
[pipe_3]
...
type=FlowChannel1Phase
orientation="0.0 -1.0 0.0"
...
[]
[pipe_4]
...
type=ElbowPipe1Phase
orientation="0.0 0.0 1.0"
....
[]
[]
```
See the second elbow needs to be [0,0,1] rather than [0,1,0] in order to get the view shown in the Paraview plot above. See the input here - https://drive.google.com/file/d/1eLtEVeUawnikdwAKmP7dP1YgnTWN9-Au/view?usp=sharing

So if for input consistency the elbows *should* follow my second diagram, then I think there must be a bug somewhere, because if I use the direction as I should, the exodus file out is incorrect.

**Reply by makeclean** [reply]
Do you follow @joshuahansel ?

**Reply by joshuahansel** [reply]
Ok, interesting, that sounds like a meshing bug in ElbowPipe1Phase. I'll look into this and get back to you.

**Reply by joshuahansel** [reply]
I've looked into this. It's very confusing. From what I can tell, here's how the points are generated for the elbow:

1. Generate an arc in the XY plane (at z=0) about the origin, from `start_angle` to `end_angle` with radius `radius`.
2. Apply rotation about the x-axis using `rotation`
3. Apply directional transformation using `orientation`
4. Apply translation using `position`

If you're really interested, see `ElbowPipe1Phase::buildMeshNodes()` for step 1 and `DiscreteLineSegmentInterface::computeRealPointFromReferencePoint()` for steps 2-4.

But I assume you want a solution :)

I think what this needs is a refactor to make the curve specification much more clear... I'll think about a good way to do this.

I apologize for all of the confusion. This is a component that we admittedly have not used very much, so we probably only tested it for a more restricted setup (like oriented in the x-direction maybe).

**Reply by joshuahansel** [reply]
I don't have any quick solution to this. I think this needs some refactoring to do right. I'm not sure if I'll get to it soon. If you want to make your own version of `ElbowPipe1Phase` that fixes it for your case, you can do that. There's not much to that class.

Here's what I'm dreaming up here: the user only specifies the ends of the flow channels to connect to, and then a smooth transition is found between these two points such that the directions at the connection points match. For example, I'm wondering if I can use Bezier curves or something to accomplish this.

**Reply by joshuahansel** [reply]
On second thought, I'd probably start with something very simple: the user specifies

- `position`: the start position
- `perpindicular_clockwise_direction`: the direction that is perpindicular to the arc, for which the arc goes in a clockwise direction from the start position
- `radius`: the radius of the arc
- `degrees`: the degrees of the arc
